### PR TITLE
Make cluster commands asynchronous

### PR DESCRIPTION
**Pull request by GGP1**
|Related issue|
|---|
| Closes https://github.com/wazuh/wazuh/issues/21862 |

## Description

Makes the  `Sendsync`, `Agent-info sync`, `Agent-groups recv`, `Agent-groups recv full` cluster commands asynchronous.

## Logs/Alerts example

Logs can be found [here](https://github.com/wazuh/wazuh/issues/21862#issuecomment-2013376134) and [here](https://github.com/wazuh/wazuh/issues/21862#issuecomment-2015514131).

<details><summary>Cluster logs after the logging changes</summary>

```console
root@wazuh-master:/# tail -f /var/ossec/logs/cluster.log
2024/04/05 18:04:37 INFO: [Master] [Local integrity] Starting.
2024/04/05 18:04:37 INFO: [Master] [Local integrity] Finished in 0.002s. Calculated metadata of 34 files.
2024/04/05 18:04:41 INFO: [Worker wazuh-worker2] [Integrity check] Starting.
2024/04/05 18:04:41 INFO: [Worker wazuh-worker2] [Integrity check] Finished in 0.003s. Received metadata of 34 files. Sync not required.
2024/04/05 18:04:41 INFO: [Worker wazuh-worker1] [Integrity check] Starting.
2024/04/05 18:04:41 INFO: [Worker wazuh-worker1] [Integrity check] Finished in 0.003s. Received metadata of 34 files. Sync not required.
2024/04/05 18:04:45 INFO: [Master] [Local integrity] Starting.
2024/04/05 18:04:45 INFO: [Master] [Local integrity] Finished in 0.002s. Calculated metadata of 34 files.
2024/04/05 18:04:45 INFO: [Worker wazuh-worker1] [Agent-info sync] Starting.
2024/04/05 18:04:45 INFO: [Worker wazuh-worker1] [Agent-info sync] Finished in 0.001s. Updated 1 chunks.
2024/04/05 18:04:46 INFO: [Master] [Local agent-groups] Starting.
2024/04/05 18:04:46 INFO: [Master] [Local agent-groups] Finished in 0.000s.
2024/04/05 18:04:46 INFO: [Worker wazuh-worker2] [Agent-groups send] Starting.
2024/04/05 18:04:46 INFO: [Worker wazuh-worker1] [Agent-groups send] Starting.
2024/04/05 18:04:46 INFO: [Worker wazuh-worker2] [Agent-groups send] Finished in 0.002s. Updated 1 chunks.
2024/04/05 18:04:48 INFO: [Worker wazuh-worker1] [Agent-groups send] Finished in 1.621s. Updated 1 chunks.
2024/04/05 18:04:50 INFO: [Worker wazuh-worker2] [Integrity check] Starting.
2024/04/05 18:04:50 INFO: [Worker wazuh-worker2] [Integrity check] Finished in 0.002s. Received metadata of 34 files. Sync not required.
2024/04/05 18:04:50 INFO: [Worker wazuh-worker1] [Integrity check] Starting.
2024/04/05 18:04:50 INFO: [Worker wazuh-worker1] [Integrity check] Finished in 0.002s. Received metadata of 34 files. Sync not required.
2024/04/05 18:04:53 INFO: [Master] [Local integrity] Starting.
2024/04/05 18:04:53 INFO: [Master] [Local integrity] Finished in 0.002s. Calculated metadata of 34 files.
2024/04/05 18:04:55 INFO: [Worker wazuh-worker1] [Agent-info sync] Starting.
2024/04/05 18:04:55 INFO: [Worker wazuh-worker1] [Agent-info sync] Finished in 0.001s. Updated 1 chunks.
2024/04/05 18:04:56 INFO: [Master] [Local agent-groups] Starting.
2024/04/05 18:04:56 INFO: [Master] [Local agent-groups] Finished in 0.000s.
2024/04/05 18:04:56 INFO: [Worker wazuh-worker2] [Agent-groups send] Starting.
2024/04/05 18:04:56 INFO: [Worker wazuh-worker1] [Agent-groups send] Starting.
2024/04/05 18:04:56 INFO: [Worker wazuh-worker2] [Agent-groups send] Finished in 0.002s. Updated 1 chunks.
2024/04/05 18:04:56 INFO: [Worker wazuh-worker1] [Agent-groups send] Finished in 0.011s. Updated 1 chunks.
2024/04/05 18:04:59 INFO: [Worker wazuh-worker2] [Integrity check] Starting.
2024/04/05 18:04:59 INFO: [Worker wazuh-worker2] [Integrity check] Finished in 0.002s. Received metadata of 34 files. Sync not required.
2024/04/05 18:04:59 INFO: [Worker wazuh-worker1] [Integrity check] Starting.
2024/04/05 18:04:59 INFO: [Worker wazuh-worker1] [Integrity check] Finished in 0.002s. Received metadata of 34 files. Sync not required.
2024/04/05 18:05:01 INFO: [Master] [Local integrity] Starting.
2024/04/05 18:05:01 INFO: [Master] [Local integrity] Finished in 0.002s. Calculated metadata of 34 files.
2024/04/05 18:05:05 INFO: [Worker wazuh-worker1] [Agent-info sync] Starting.
2024/04/05 18:05:05 INFO: [Worker wazuh-worker1] [Agent-info sync] Finished in 0.001s. Updated 1 chunks.
2024/04/05 18:05:06 INFO: [Master] [Local agent-groups] Starting.
2024/04/05 18:05:06 INFO: [Master] [Local agent-groups] Finished in 0.000s.
2024/04/05 18:05:06 INFO: [Worker wazuh-worker2] [Agent-groups send] Starting.
2024/04/05 18:05:06 INFO: [Worker wazuh-worker1] [Agent-groups send] Starting.
2024/04/05 18:05:06 INFO: [Worker wazuh-worker2] [Agent-groups send] Finished in 0.003s. Updated 1 chunks.
2024/04/05 18:05:06 INFO: [Worker wazuh-worker1] [Agent-groups send] Finished in 0.009s. Updated 1 chunks.
2024/04/05 18:05:08 INFO: [Worker wazuh-worker2] [Integrity check] Starting.
2024/04/05 18:05:08 INFO: [Worker wazuh-worker2] [Integrity check] Finished in 0.003s. Received metadata of 34 files. Sync not required.
2024/04/05 18:05:08 INFO: [Worker wazuh-worker1] [Integrity check] Starting.
2024/04/05 18:05:08 INFO: [Worker wazuh-worker1] [Integrity check] Finished in 0.004s. Received metadata of 34 files. Sync not required.
2024/04/05 18:05:09 INFO: [Master] [Local integrity] Starting.
2024/04/05 18:05:09 INFO: [Master] [Local integrity] Finished in 0.002s. Calculated metadata of 34 files.
...
```

</details>

### CODE REVIEW

**Review by fdalmaup — CHANGES_REQUESTED** [review]
Changes look good and the manual tests do not seem to have issues. Nevertheless, we should run [related system](https://github.com/wazuh/wazuh-qa/tree/master/tests/system/test_cluster) and [cluster performance](https://github.com/wazuh/wazuh-qa/tree/master/tests/performance/test_cluster) tests as part of the verification process.

**Review by fdalmaup — APPROVED** [review]
LGTM! The required tests were run [here](https://github.com/wazuh/wazuh/issues/21862#issuecomment-2034832186).

**Comment by Selutario** [comment]
It has been decided not to publish a 4.8.2 patch for now. I'm moving this issue to on hold for it to be rebased to 4.9.0.

**Review by Selutario — CHANGES_REQUESTED** [review]
It seems we will need to update the cluster documentation too, now multiprocessing is used in less places:
- https://documentation.wazuh.com/current/development/wazuh-cluster.html#multiprocessing

File: framework/wazuh/core/cluster/common.py [hunk 1]
```diff
@@ -593,13 +593,12 @@ async def update_chunks_wdb(self, data: dict, info_type: str, logger: logging.Lo
             Dict containing number of updated chunks, error messages (if any) and time spent.
         """
         try:
-            result = await cluster.run_in_pool(self.loop, self.server.task_pool, send_data_to_wdb, data,
-                                               timeout, info_type=info_type)
+            result = await send_data_to_wdb(data, timeout, info_type)
```
[the remark this answers is not in the document]
Comment (GGP1) [reply framework/wazuh/core/cluster/common.py:596]: I agree, since it will execute asynchronously from now on I think we can get rid of the timeout.

**Review by Selutario — APPROVED** [review]
LGTM, nice job @GGP1. These changes affect key parts of the cluster, we will have to pay special attention during the release testing to detect any strange behavior that we may be overlooking now.

**Pull request merged**